FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+      IREmit->_StoreContext(1, GPRClass, GetConstant(Mask), offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+    } else if (Mask != 0) {
+      if (std::popcount(Mask) == 1) {
+        uint8_t BitIdx = __builtin_ctz(Mask);
+        SetX87ValidTag(GetOffsetTopWithCache_Slow(BitIdx), true);
+      } else {
+        // perform a rotate right on mask by top
+        // since we must operate on 32bits as a minimum:
+        // ror (mask, top) = ((M << 8) | M) >> TOP
+        auto* TopValue = GetTopWithCache_Slow();
+        Ref RotAmount = IREmit->_Sub(OpSize::i32Bit, GetConstant(8), TopValue);
+        Ref RMask = IREmit->_Or(OpSize::i32Bit, MaskC, IREmit->_Lshl(OpSize::i32Bit, MaskC, GetConstant(8)));
+        RMask = IREmit->_Lshr(OpSize::i32Bit, RMask, RotAmount);
+        Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+        Ref NewAbridgedFTW = IREmit->_Or(OpSize::i32Bit, AbridgedFTW, RMask);
+        IREmit->_StoreContext(1, GPRClass, NewAbridgedFTW, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+      }
+    } else {
+      LogMan::Msg::DFmt("No valid tags written");
+    }
+  }
+  { // Set invalid tags
+    uint8_t Mask = StackData.getInvalidMask();
+    Ref MaskC = GetConstant(Mask);
+    LogMan::Msg::DFmt("Writing invalid tags 0x{:02x}", Mask);
+    if (Mask == 0xff) {
+      IREmit->_StoreContext(1, GPRClass, GetConstant(0), offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+    } else if (Mask != 0) {
+      if (std::popcount(Mask)) {
+        uint8_t BitIdx = __builtin_ctz(Mask);
+        SetX87ValidTag(GetOffsetTopWithCache_Slow(BitIdx), false);
+      } else {
+        // Same rotate right as above but this time on the invalid mask
+        auto* TopValue = GetTopWithCache_Slow();
+        Ref RotAmount = IREmit->_Sub(OpSize::i32Bit, GetConstant(8), TopValue);
+        Ref RMask = IREmit->_Or(OpSize::i32Bit, MaskC, IREmit->_Lshl(OpSize::i32Bit, MaskC, GetConstant(8)));
+        RMask = IREmit->_Lshr(OpSize::i32Bit, RMask, RotAmount);
+        Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+        Ref NewAbridgedFTW = IREmit->_Andn(OpSize::i32Bit, AbridgedFTW, RMask);
+        IREmit->_StoreContext(1, GPRClass, NewAbridgedFTW, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+      }
+    } else {
+      LogMan::Msg::DFmt("No invalid tags written");
+    }
+  }
+  return TopValue;
+}
+
+std::tuple<Ref, Ref> X87StackOptimization::SplitF64SigExp(Ref Node) {
+  Ref Gpr = IREmit->_VExtractToGPR(8, 8, Node, 0);
+
+  Ref Exp = IREmit->_And(OpSize::i64Bit, Gpr, GetConstant(0x7ff0000000000000LL));
+  Exp = IREmit->_Lshr(OpSize::i64Bit, Exp, GetConstant(52));
+  Exp = IREmit->_Sub(OpSize::i64Bit, Exp, GetConstant(1023));
+  Exp = IREmit->_Float_FromGPR_S(8, 8, Exp);
+  Ref Sig = IREmit->_And(OpSize::i64Bit, Gpr, GetConstant(0x800fffffffffffffLL));
+  Sig = IREmit->_Or(OpSize::i64Bit, Sig, GetConstant(0x3ff0000000000000LL));
```

Comment:
We know that none of the constants in this function will hit this pass's constant cache, might as well use _Constant() directly

Reply:
The thought was that if the Constant cache is increased, to a certain point then this would be automatically caught in the future. And if it doesn't then I don't expect it to have performance impact.

Reply:
I guess this duplicates the thread of "why do we even have a constant cache here, does this actually make a ton of sense?" then..